src/gleam/http.gleam

```diff
@@ -52,3 +61,158 @@ pub fn method_to_string(method) {
 
 pub external fn method_from_erlang(anything) -> Result(Method, Nil) =
   "gleam_http_native" "method_from_erlang"
+
+/// Data type containing an attribute value pair that constitute an HTTP header.
+pub type Header =
+  tuple(String, String)
+
+/// A general data type that can be be parameterised as a `Request` or `Response`
+pub opaque type Message(head, body) {
+  Message(head: head, headers: List(Header), body: body)
+}
+
+/// Data structure containing the deserialized components of an HTTP request start line.
+pub opaque type RequestHead {
+  RequestHead(
+    method: Method,
+    host: String,
+    port: Option(Int),
+    path: String,
+    query: Option(String),
+  )
+}
+
+/// Data structure containing the deserialized components of an HTTP response status line.
+pub opaque type ResponseHead {
+  ResponseHead(status: Int)
+}
+
+/// HTTP message type parameterised as a Request.
+pub type Request(body) =
+  Message(RequestHead, body)
+
+/// HTTP message type parameterised as a Response.
+pub type Response(body) =
+  Message(ResponseHead, body)
+
+/// Construct a `Request` type.
+///
+/// The provided `uri_string` should be an absolure uri, including the scheme and host.
+/// The body type of the returned request is `Nil`,
+/// and should be set with a call to `set_body`.
+pub fn request(method: Method, uri_string: String) -> Request(Nil) {
+  let Ok(
```

Comment:
This line will result in a crash. It should definetly be possible to create a request without having to handle a result return value. I think this fn should just accept port/path/query etc as arguments.

There may be a future function `from_string(uri: String)` or `from_uri(uri: Uri)` for creating request but they can come later.